nvflare/fuel/f3/drivers/base_driver.py

```diff
@@ -0,0 +1,66 @@
+# Copyright (c) 2021-2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+import threading
+from abc import ABC
+from typing import Dict, Optional
+
+from nvflare.fuel.f3.connection import Connection, ConnState
+from nvflare.fuel.f3.drivers.driver import ConnectorInfo, Driver
+
+log = logging.getLogger(__name__)
+
+
+class BaseDriver(Driver, ABC):
+    """Common base class for all drivers
+    It contains all the common connection management code
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.connections: Dict[str, Connection] = {}
+        self.connector: Optional[ConnectorInfo] = None
+        self.conn_lock = threading.Lock()
+
+    def add_connection(self, conn: Connection):
+        with self.conn_lock:
+            self.connections[conn.name] = conn
+
+        conn.state = ConnState.CONNECTED
+        self._notify_monitor(conn)
+
+        log.debug(f"Connection created: {self.get_name()}:{conn}")
+
+    def close_connection(self, conn: Connection):
+        log.debug(f"Connection removed: {self.get_name()}:{conn}")
+
+        conn.state = ConnState.CLOSED
+        self._notify_monitor(conn)
+
+        with self.conn_lock:
+            if not self.connections.pop(conn.name, None):
+                log.debug(f"{conn.name} is already removed from driver")
+
+    def close_all(self):
+        with self.conn_lock:
+            for name in sorted(self.connections.keys()):
+                conn = self.connections[name]
+                log.debug(f"Closing connection: {self.get_name()}:{conn}")
+                conn.close()
+
+    def _notify_monitor(self, conn: Connection):
+        if not self.conn_monitor:
+            log.error(f"Connection monitor not registered for driver {self.get_name()}")
```

Comment:
if this is a programming error, should we raise Exception?

Reply:
It's an error for production use. In tests, you may not care about the connection state. So it's better to log an error.

Reply:
But we should assume users are not doing testing right?

Reply:
Not for users, for developers. The error and exception have the same effect, they result in errors in the log.

Reply:
I agree "log error" or "log exception" is similar.

But "log error" is different from "raising Exception"?

Reply:
The only difference is that raising an exception will cause the connection to end and the error is logged once. While in current implementation, the connection still works and the error will be logged twice when state changes to CONNECTED and DISCONNECTED. I think latter is better.

Reply:
If the driver is written wrong / programming error, we should just ends the program?

Reply:
Let's talk about this and deal with it in another PR.